File: java/_20_Valid_Parentheses.java
import java.util.Stack;

public class _20_Valid_Parentheses {
	 public boolean isValid(String s) {
			if(s.length()==0)
				return true;
			if(s.length()%2==1)
				return false;
			Stack<String>stack=new Stack<>();
			if(s.charAt(0)==')'||s.charAt(0)==']'||s.charAt(0)=='}')
				return false;
			stack.push(s.substring(0, 1));
			for(int i=1;i<s.length();i++) {
				if(s.charAt(i)=='('||s.charAt(i)=='['||s.charAt(i)=='{') {
					stack.push(s.substring(i, i+1));
				}
				else {
					String tmp=stack.peek();
					if(tmp.equals("(")&&s.substring(i,i+1).equals(")"))
						stack.pop();
					else if(tmp.equals("[")&&s.substring(i,i+1).equals("]"))
						stack.pop();
					else if(tmp.equals("{")&&s.substring(i,i+1).equals("}"))
						stack.pop();
					else
						return false;
				}
			}
			if(stack.isEmpty())
				return true;
			return false;
	 }
}

File: cpp/20.Valid Parentheses.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char>_stack;
        for (int i = 0; i < s.size(); i++){
            if (s[i] == '(' || s[i] == '[' || s[i] == '{'){
                _stack.push(s[i]);
            }
            else{
                if (_stack.empty())
                    return false;
                char pop_char = _stack.top();
                _stack.pop();
                if (s[i] == ')' && pop_char != '(')
                    return false;
                if (s[i] == ']' && pop_char != '[')
                    return false;
                if (s[i] == '}' && pop_char != '{')
                    return false;
            }
        }
        if (_stack.empty())
            return true;
        else
            return false;
    }
};

